ethtool_link_settings.go

```diff
@@ -0,0 +1,327 @@
+		return nil, fmt.Errorf("ETHTOOL_GLINKSETTINGS ioctl failed: %w", err)
+	}
+
+	// If we reach here, it means attemptFallback is true
+	if attemptFallback {
+		// Fallback to ETHTOOL_GSET using e.CmdGet
+		var cmd EthtoolCmd
+		_, errGet := e.CmdGet(&cmd, intf)
+		if errGet != nil {
+			return nil, fmt.Errorf("ETHTOOL_GLINKSETTINGS failed (%s), fallback ETHTOOL_GSET (CmdGet) also failed: %w", fallbackReason, errGet)
+		}
+		results := convertCmdToLinkSettings(&cmd)
+		results.Source = "GSET"
+		return results, nil
+	}
+
+	// Should be unreachable if logic is exhaustive
+	return nil, fmt.Errorf("unexpected state at end of GetLinkSettings")
+}
+
+// SetLinkSettings applies link settings, determining whether to use ETHTOOL_SLINKSETTINGS or ETHTOOL_SSET.
+func (e *Ethtool) SetLinkSettings(intf string, settings *LinkSettings) error {
+	var checkReq ethtoolLinkSettingsRequest
+	checkReq.Settings.Cmd = ETHTOOL_GLINKSETTINGS
+	checkReq.Settings.LinkModeMasksNwords = int8(MAX_LINK_MODE_MASK_NWORDS)
+
+	errGLinkSettings := e.ioctl(intf, uintptr(unsafe.Pointer(&checkReq)))
+	canUseGLinkSettings := false
+	nwords := 0
+
+	if errGLinkSettings == nil {
+		nwords = int(checkReq.Settings.LinkModeMasksNwords)
+		if nwords <= 0 || nwords > MAX_LINK_MODE_MASK_NWORDS {
+			return fmt.Errorf("ETHTOOL_GLINKSETTINGS check succeeded but returned invalid nwords: %d", nwords)
+		}
+		canUseGLinkSettings = true
+	} else if errno, ok := errGLinkSettings.(syscall.Errno); ok && errno == unix.EOPNOTSUPP {
+		canUseGLinkSettings = false
+	} else {
+		return fmt.Errorf("checking support via ETHTOOL_GLINKSETTINGS failed: %w", errGLinkSettings)
+	}
+
+	if canUseGLinkSettings {
+		var setReq ethtoolLinkSettingsRequest
+		if 3*nwords > len(setReq.Masks) {
+			return fmt.Errorf("internal error: required nwords (%d) exceeds allocated buffer (%d)", nwords, MAX_LINK_MODE_MASK_NWORDS)
+		}
+		setReq.Settings.Cmd = ETHTOOL_SLINKSETTINGS
+		setReq.Settings.Speed = settings.Speed
+		setReq.Settings.Duplex = settings.Duplex
+		setReq.Settings.Port = settings.Port
+		setReq.Settings.PhyAddress = settings.PhyAddress
+		setReq.Settings.Autoneg = settings.Autoneg
+		setReq.Settings.EthTpMdixCtrl = settings.EthTpMdixCtrl
+		setReq.Settings.MasterSlaveCfg = settings.MasterSlaveCfg
+		setReq.Settings.LinkModeMasksNwords = int8(nwords)
+
+		advertisingMask := buildLinkModeMask(settings.AdvertisingLinkModes, nwords)
+		if len(advertisingMask) != nwords {
+			return fmt.Errorf("failed to build advertising mask with correct size (%d != %d)", len(advertisingMask), nwords)
+		}
+		copy(setReq.Masks[nwords:2*nwords], advertisingMask)
+		zeroMaskSupported := make([]uint32, nwords)
+		zeroMaskLp := make([]uint32, nwords)
+		copy(setReq.Masks[0*nwords:1*nwords], zeroMaskSupported)
+		copy(setReq.Masks[2*nwords:3*nwords], zeroMaskLp)
+
+		if err := e.ioctl(intf, uintptr(unsafe.Pointer(&setReq))); err != nil {
+			return fmt.Errorf("ETHTOOL_SLINKSETTINGS ioctl failed: %w", err)
+		}
+		return nil
+
+	} else {
```

Comment:
nit: this `else` line is probably not necessary as you have a `return` in the `if`